### ISSUE TITLE
J9 crashes on a correct test case

### ISSUE THREAD

**Issue by JavaTailor** [eclipse-openj9/openj9#14737]
## Affected versions

We found a test case with crash problems. To facilitate analysis, we simplified the test case and the simplified class file can ben found at attachment.

Windows 10:

```
Microsoft Windows 10 Professional
10.0.19044 Build 19044
AMD Ryzen 5 5600G with Radeon Graphics 3.90 GHz
Memory 32 GB
```

Java -version output under Windows 10

```
openjdk version "1.8.0_312"
IBM Semeru Runtime Open Edition (build 1.8.0_312-b07)
Eclipse OpenJ9 VM (build openj9-0.29.0, JRE 1.8.0 Windows 10 amd64-64-Bit Compressed References 20211022_236 (JIT enabled, AOT enabled)
OpenJ9   - e1e72c497
OMR      - 299b6a2d2
JCL      - 8860d39588 based on jdk8u312-b07
```

```
openjdk version "11.0.13" 2021-10-19
IBM Semeru Runtime Open Edition 11.0.13.0 (build 11.0.13+8)
Eclipse OpenJ9 VM 11.0.13.0 (build openj9-0.29.0, JRE 11 Windows 10 amd64-64-Bit Compressed References 20211022_218 (JIT enabled, AOT enabled)
OpenJ9   - e1e72c497
OMR      - 299b6a2d2
JCL      - 2d83aa3b76 based on jdk-11.0.13+8)
```

```
openjdk version "16.0.2" 2021-07-20
IBM Semeru Runtime Open Edition 16.0.2.0 (build 16.0.2+7)
Eclipse OpenJ9 VM 16.0.2.0 (build openj9-0.27.0, JRE 16 Windows 10 amd64-64-Bit Compressed References 20210730_69 (JIT enabled, AOT enabled)
OpenJ9   - 1851b0074
OMR      - 9db1c870d
JCL      - 34df42439f3 based on jdk-16.0.2+7)
```

## Problem summary

Given the simplified test case, Hotspot executes fine, but J9 crashes along with three crash log file. Since the crash log file is too large, here we only show the crash message we got from the console output.

```
Unhandled exception
Type=Segmentation error vmState=0x00000000
Windows_ExceptionCode=c0000005 J9Generic_Signal=00000004 ExceptionAddress=00000000000001E8 ContextFlags=0010005f
Handler1=00007FF8DCBFD500 Handler2=00007FF8DB2E93A0 DEPAddress=00000000000001E8
RDI=0000000000056C00 RSI=0000000000000014 RAX=00000007FFF6F830 RBX=0000000000000013
RCX=0000000000000014 RDX=00000007FFF6F620 R8=0000000000000000 R9=0000000000000013
R10=0000000000000013 R11=0000000000000013 R12=0000000000000013 R13=00000007FFF6F620
R14=0000000000000010 R15=00007FF7102AFF28
RIP=00000000000001E8 RSP=00000000000FC228 RBP=000000000001E400 EFLAGS=0000000000010246
FS=0053 ES=002B DS=002B
XMM0 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM1 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM2 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM3 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM4 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM5 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM6 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM7 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM8 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM9 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM10 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM11 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM12 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM13 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM14 0000000000000000 (f: 0.000000, d: 0.000000e+000)
XMM15 0000000000000000 (f: 0.000000, d: 0.000000e+000)
Module=D:\,Target=2_90_20211022_236 (Windows 10 10.0 build 19044)
CPU=amd64 (12 logical CPUs) (0x7d6a62000 RAM)
----------- Stack Backtrace -----------
(0x00000000000001E8)
(0x00007FF448C88BDA)
(0x4034000000000000)
(0x000000060A5B3FE0)
(0x00000007FFF6F780)
---------------------------------------
```

## Attachment
[TestCase3.zip](https://github.com/eclipse-openj9/openj9/files/8282461/TestCase3.zip)

**Comment by pshipton** [comment]
The crash doesn't occur with -Xint, assigned to JIT. @0xdaryl @mstoodle fyi

Seems to be cross platform. I tried xlinux, zlinux.

I note this affects Java 18 as well, but I've assigned it to the 0.32 milestone.

**Comment by gacholio** [comment]
Crash is due to bad virtual call from:
```
J9JITExceptionTable at 0x00007F2148F6F878 {
    struct J9UTF8* className = !j9utf8 0x00007F216046A002   // TestCase3_1
    struct J9UTF8* methodName = !j9utf8 0x00007F216046A01E   // main
    struct J9UTF8* methodSignature = !j9utf8 0x00007F2110028E84   // ([Ljava/lang/String;)V
    struct J9ConstantPool* constantPool = !j9constantpool 0x000000000016E2C0 
    struct J9Method* ramMethod = !j9method 0x000000000016E3C8   // TestCase3_1.main([Ljava/lang/String;)V
    UDATA parm.startPC = 0x00007F21498069D4;
    UDATA parm.endWarmPC = 0x00007F2149806BD2;
    UDATA parm.startColdPC = 0x0000000000000000;
    UDATA parm.endPC = 0x00007F2149806BD2;
    UDATA parm.totalFrameSize = 0x0000000000000005;
    I_16 parm.slots = 0x0001;
    I_16 parm.scalarTempSlots = 0x0000;
    I_16 parm.objectTempSlots = 0x0000;
    U_16 parm.prologuePushes = 0x0002;
    I_16 parm.tempOffset = 0x0000;
    U_16 parm.numExcptionRanges = 0x0000;
    I_32 parm.size = 0x0000015E;
    UDATA parm.registerSaveDescription = 0x0000000000000202;
    void* gcStackAtlas = !void 0x00007F2148F6F961 
    void* inlinedCalls = !void 0x00007F2148F6F950 
    void* bodyInfo = !void 0x0000000000000000 
    struct J9JITExceptionTable* nextMethod = !j9jitexceptiontable 0x0000000000000000 
    struct J9JITExceptionTable* prevMethod = !j9jitexceptiontable 0x0000000000000000 
    void* debugSlot1 = !void 0x0000000000000000 
    void* debugSlot2 = !void 0x0000000000000000 
    void* osrInfo = !void 0x0000000000000000 
    void* runtimeAssumptionList = !void 0x00007F215B3D20F0 
    I_32 parm.hotness = 0x00000003;
    UDATA parm.codeCacheAlloc = 0x00007F21498069D0;
    void* gpuCode = !void 0x0000000000000000 
    void* riData = !void 0x0000000000000000 
}

   0x7f2149806ae5:	movabs $0x4ea00,%rdi
   0x7f2149806aef:	lea    0x0(,%rdi,1),%edi
   0x7f2149806af6:	callq  *-0x40(%rdi)
```
The value loaded into RDI is a J9Class pointer not an object:
```
J9ArrayClass at 0x4ea00 {
  Fields for J9ArrayClass:
	0x0: UDATA eyecatcher = 0x0000000099669966 (2573637990)
	0x8: struct J9ROMClass * romClass = !j9romclass 0x00007F215FDF6670
	0x10: struct J9Class ** superclasses = !j9x 0x000000000004BB40
	0x18: UDATA classDepthAndFlags = 0x00000000800D0001 (2148335617)
	0x20: U32 classDepthWithFlags = 0x00000000 (0)
	0x24: U32 classFlags = 0x00000000 (0)
	0x28: struct J9ClassLoader * classLoader = !j9classloader 0x00007F21600854E8
	0x30: j9object_t classObject = !j9object 0x00000000E0000CC0 // java/lang/Class
	0x38: volatile UDATA initializeStatus = 0x0000000000000001 (1)
	0x40: struct J9Class * leafComponentType = !j9class 0x0000000000048F00 // java/lang/String
	0x48: UDATA arity = 0x0000000000000001 (1)
	0x50: struct J9Class * arrayClass = !j9class 0x0000000000000000
	0x58: struct J9Class * componentType = !j9class 0x0000000000048F00 // java/lang/String
	0x60: struct J9ITable * lastITable = !j9itable 0x0000000000044088
	0x68: UDATA * instanceDescription = !j9x 0x0000000000000001
	0x70: UDATA * instanceLeafDescription = !j9x 0x0000000000000001
	0x78: UDATA instanceHotFieldDescription = 0x0000000000000000 (0)
	0x80: UDATA selfReferencingField1 = 0x0000000000000000 (0)
	0x88: UDATA selfReferencingField2 = 0x0000000000000000 (0)
	0x90: struct J9Method * initializerCache = !j9method 0x0000000000000000
	0x98: UDATA romableAotITable = 0x00007F215DF9B880 (139781287295104)
	0xa0: UDATA packageID = 0x00007F21600854E8 (139781321807080)
	0xa8: struct J9Module * module = !j9module 0x0000000000000000
	0xb0: struct J9Class * subclassTraversalLink = !j9class 0x000000000004DE00 // sun/misc/Unsafe
	0xb8: struct J9Class * subclassTraversalReverseLink = !j9class 0x000000000004F400 // sun/reflect/Reflection
	0xc0: void ** iTable = !j9x 0x0000000000044088
	0xc8: UDATA castClassCache = 0x000000000016E101 (1499393)
	0xd0: void ** jniIDs = !j9x 0x0000000000000000
	0xd8: UDATA lockOffset = 0xFFFFFFFFFFFFFFFF (-1)
	0xe0: U32 paddingForGLRCounters = 0x00000000 (0)
	0xe4: U16 reservedCounter = 0x0000 (0)
	0xe6: U16 cancelCounter = 0x0000 (0)
	0xe8: UDATA newInstanceCount = 0x00000000000007D0 (2000)
	0xf0: IDATA backfillOffset = 0x0000000000000004 (4)
	0xf8: struct J9Class * replacedClass = !j9class 0x0000000000000000
	0x100: UDATA finalizeLinkOffset = 0x0000000000000000 (0)
	0x108: struct J9Class * nextClassInSegment = !j9class 0x000000000004DE00 // sun/misc/Unsafe
	0x110: UDATA * ramConstantPool = !j9x 0x0000000000000000
	0x118: j9object_t * callSites = !j9x 0x0000000000000000
	0x120: j9object_t * methodTypes = !j9x 0x0000000000000000
	0x128: j9object_t * varHandleMethodTypes = !j9x 0x0000000000000000
	0x130: struct J9VMCustomSpinOptions * customSpinOption = !j9vmcustomspinoptions 0x0000000000000000
	0x138: struct J9Method ** staticSplitMethodTable = !j9x 0x0000000000000000
	0x140: struct J9Method ** specialSplitMethodTable = !j9x 0x0000000000000000
	0x148: struct J9JITExceptionTable * jitMetaDataList = !j9jitexceptiontable 0x0000000000000000
	0x150: struct J9Class * gcLink = !j9class 0x0000000000000000
	0x158: struct J9Class * hostClass = !j9class 0x000000000004EA00 // [Ljava/lang/String;
	0x160: UDATA flattenedElementSize = 0x0000000000000004 (4)
	0x168: struct J9ClassHotFieldsInfo * hotFieldsInfo = !j9classhotfieldsinfo 0x0000000000000000
}
```

**Comment by gacholio** [comment]
Reproducible with `-Xjit:verbose,limit={TestCase3_1.main*}`. Looks like it might be a DLT issue:
```
+ (warm) TestCase3_1.main([Ljava/lang/String;)V @ 00007F4DC3200060-00007F4DC3200451 OrdinaryMethod - Q_SZ=1 Q_SZI=0 QW=13 j9m=000000000016E3C8 bcsz=189 GCR compThreadID=0 CpuLoad=116%(29%avg) JvmCpu=83%
+ (hot) TestCase3_1.main([Ljava/lang/String;)V @ 00007F4DC3200474-00007F4DC3200672 MethodInProgress - Q_SZ=0 Q_SZI=0 QW=1 j9m=000000000016E3C8 bcsz=189 DLT@119 compThreadID=0 CpuLoad=116%(29%avg) JvmCpu=83%
```

**Comment by 0xdaryl** [comment]
This is almost certainly a dup of #14716.  The crash in this issue is resolved when the proposed fix for #14716 is applied.

**Comment by pshipton** [comment]
I'm closing it as a dup.
